lib/util/zip.ex

```diff
@@ -0,0 +1,139 @@
+# Copyright(c) 2015-2019 ACCESS CO., LTD. All rights reserved.
+
+use Croma
+
+defmodule Antikythera.Zip do
+  @moduledoc """
+  Wrapper module for `zip` command.
+
+  For the consistency in working with antikythera and other gears, scope of this module is limited under a temporary directory reserved by `Antikythera.Tmpdir.make/2`.
+
+  Functions only accept absolute paths for both source and resulting archive.
+  """
+
+  alias Croma.Result, as: R
+  alias Antikythera.Context
+  alias Antikythera.ExecutorPool.Id, as: EPoolId
+  alias AntikytheraCore.TmpdirTracker
+
+  @typep opts :: {:encryption, boolean} | {:password, String.t}
+
+  defmodule FileName do
+    use Croma.SubtypeOfString, pattern: ~r/^(?!.*\/\.{0,2}\z).*\z/
+  end
+
+  @doc """
+  Creates a ZIP file.
+
+  Encryption using a password is supported in which `encryption` option is set `true`.
+
+  ## Example
+    Tmpdir.make(context, fn tmpdir ->
+      src_path = tmpdir <> "/src.txt"
+      zip_path = tmpdir <> "/archive.zip"
+      File.write!(src_path, "text")
+      Antikythera.Zip.zip(context, zip_path, src_path, [encryption: true, password: "password"])
+      |> case do
+        {:ok, archive_path} ->
+          ...
+      end
+      ...
+    end)
+  """
+  defun zip(context_or_epool_id :: v[EPoolId.t | Context.t],
+            zip_raw_path        :: v[FileName.t],
+            src_raw_path        :: v[String.t],
+            opts                :: v[list(opts)] \\ []) :: R.t(Path.t) do
+    epool_id = extract_epool_id(context_or_epool_id)
+    with {:ok, tmpdir} <- TmpdirTracker.get(epool_id),
+         :ok           <- reject_existing_dir(zip_raw_path),
+         :ok           <- validate_suffix(src_raw_path),
+         zip_path      <- Path.expand(zip_raw_path),
+         src_path      <- Path.expand(src_raw_path),
+         :ok           <- validate_within_tmpdir(zip_path, tmpdir),
+         :ok           <- validate_within_tmpdir(src_path, tmpdir),
+         :ok           <- validate_path_exists(src_path),
+         :ok           <- ensure_dir_exists(zip_path, tmpdir),
+         {:ok, args}   <- opts |> Map.new() |> extract_zip_args(),
+         :ok           <- try_zip_cmd(args ++ [zip_path, src_path]) do
+      if zip_path |> Path.basename() |> String.contains?(".") do
+        {:ok, zip_path}
+      else
+        {:ok, zip_path <> ".zip"}
+      end
+    end
+  end
+
+  defp extract_epool_id(%Context{executor_pool_id: epool_id}), do: epool_id
+  defp extract_epool_id(epool_id),                             do: epool_id
+
+  defunp validate_suffix(path :: v[String.t]) :: :ok | {:error, tuple} do
```

Comment:
It's a trivial point, but let's change the order of function definition of `validate_suffix/1` and `reject_existing_dir/1`. (Or the order in the `with` expression)

Reply:
Changed `defunp` order in 0611d0b .